### PR TITLE
Add support for Azure Managed Prometheus in Clusterloader2

### PR DESCRIPTION
**Pull request by pink-milk**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR extends support to use [Azure Managed Prometheus](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-overview) instead of the in-cluster Prometheus instance. This can be enabled through `prometheusClient: managed` parameter in the Prometheus-based measurements' params in the test scenario config (currently available for limited number of providers). 

This change enables CL2 to collect Prometheus measurements on AKS managed clusters, a capability that was previously unavailable.

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:2]:
```suggestion
Copyright 2024 The Kubernetes Authors.
```

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+// aksManagedPrometheusClient talks to the Azure Managed Prometheus.
+// This only works if the cluster is hosted in AKS.
+// Overview: https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-overview.
+// Query API: https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-api-promql .
+type aksManagedPrometheusClient struct {
+	uri   			string
+	headers    		map[string]string
+}
+
+func (mpc *aksManagedPrometheusClient) Query(query string, queryTime time.Time) ([]byte, error) {
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:41]: Can you add a header comment for this?

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	req, err := http.NewRequest("GET", mpc.uri, nil)
+    if err != nil {
+        return nil, err
+    }
+
+	// Set query parameters
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:47]: Seems like something weird is going on with indentation. Are you using all tabs?

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    q := req.URL.Query()
+    q.Add("query", query)
+	q.Add("time", queryTime.Format(time.RFC3339))
+    req.URL.RawQuery = q.Encode()
+
+	// Set headers
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:53]: Ditto

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    for key, value := range mpc.headers {
+        req.Header.Add(key, value)
+    }
+
+	// Execute the request
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:58]: Ditto

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		return nil, err
+	}
+	if statusCode := resp.StatusCode; statusCode > 299 {
+		return respBody, fmt.Errorf("response failed with status code %d", statusCode)
+	}
+	return respBody, nil
+}
+
+func getPromQLToken(credential *azidentity.DefaultAzureCredential) (string, error) {
+    tokenRequestContext := []string{"https://prometheus.monitor.azure.com/.default"}
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:78]: Maybe save this as a const?

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        return "Failed to get Azure Prometheus Token", err
+    }
+    token := accessToken.Token
+    return token, nil
+}
+
+// NewAKSManagedPrometheusClient returns instance of aksManagedPrometheusClient 
+// which is configured to query Azure Managed Prometheus endpoint.
+func NewAKSManagedPrometheusClient() (Client, error) {
+	credential, err := azidentity.NewDefaultAzureCredential(&azidentity.DefaultAzureCredentialOptions{})
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:93]: Indentation issue

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    accessToken, err := credential.GetToken(ctx, policy.TokenRequestOptions{Scopes: tokenRequestContext})
+    if err != nil {
+        return "Failed to get Azure Prometheus Token", err
+    }
+    token := accessToken.Token
+    return token, nil
+}
+
+// NewAKSManagedPrometheusClient returns instance of aksManagedPrometheusClient
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:90]:
```suggestion
// NewAKSManagedPrometheusClient returns an instance of aksManagedPrometheusClient 
```

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    accessToken, err := credential.GetToken(ctx, policy.TokenRequestOptions{Scopes: tokenRequestContext})
+    if err != nil {
+        return "Failed to get Azure Prometheus Token", err
+    }
+    token := accessToken.Token
+    return token, nil
+}
+
+// NewAKSManagedPrometheusClient returns instance of aksManagedPrometheusClient 
+// which is configured to query Azure Managed Prometheus endpoint.
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:91]:
```suggestion
// which is configured to query the Azure Managed Prometheus endpoint.
```

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+func NewAKSManagedPrometheusClient() (Client, error) {
+	credential, err := azidentity.NewDefaultAzureCredential(&azidentity.DefaultAzureCredentialOptions{})
+	if err != nil {
+		return nil, err
+	}
+	promQLToken, err := getPromQLToken(credential)
+    if err != nil {
+        fmt.Println("Failed to get Azure Prometheus :", err)
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:99]: We don't want to use `fmt.Println` in Kubernetes controllers, instead use `klog.V(0).Infof()` where 0 is the verbosity level and can be set to any int. For errors, we want to use `klog.Errorf`

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func getPromQLToken(credential *azidentity.DefaultAzureCredential) (string, error) {
+    tokenRequestContext := []string{"https://prometheus.monitor.azure.com/.default"}
+    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+    defer cancel()
+
+    accessToken, err := credential.GetToken(ctx, policy.TokenRequestOptions{Scopes: tokenRequestContext})
+    if err != nil {
+        return "Failed to get Azure Prometheus Token", err
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:84]:
```suggestion
        return "", err
```

Typically, we leave the return value as empty or nil when there is an error.

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	"net/http"
+	
+	"time"
+	"os"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+)
+
+// aksManagedPrometheusClient talks to the Azure Managed Prometheus.
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:32]:
```suggestion
// aksManagedPrometheusClient is a client for Azure Managed Prometheus.
```

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+// This only works if the cluster is hosted in AKS.
+// Overview: https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-overview.
+// Query API: https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-api-promql .
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:35]:
```suggestion
// Query API: https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-api-promql.
```

**Review by Jont828 — COMMENTED** [review]
Congrats on your first contribution! There's just some small details, otherwise LGTM.

File: clusterloader2/pkg/prometheus/clients/aks_managed.go [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	token := accessToken.Token
+	return token, nil
+}
+
+// NewAKSManagedPrometheusClient returns an instance of aksManagedPrometheusClient
+// which is configured to query the Azure Managed Prometheus endpoint.
+func NewAKSManagedPrometheusClient() (Client, error) {
+	// Check if AZURE_QUERY_ENDPOINT is set
+	endpoint := os.Getenv("AZURE_QUERY_ENDPOINT")
```
Comment (Jont828) [inline clusterloader2/pkg/prometheus/clients/aks_managed.go:103]: Could we save the var name as a const?

**Review by Jont828 — COMMENTED** [review]
LGTM pending one comment!

**Comment by Jont828** [comment]
Nice work!
/lgtm

**Comment by marseel** [comment]
lgtm, but could you squash commits and rebase instead of merging the master branch please ?

**Pull request merged**